pydantic_ai_slim/pydantic_ai/models/xai.py

```diff
@@ -0,0 +1,868 @@
+                parts.append(
+                    BuiltinToolReturnPart(
+                        tool_name=tool_call.function.name,
+                        content={'status': 'completed'},
+                        tool_call_id=tool_call.id,
+                        provider_name=self.system,
+                    )
+                )
+            else:
+                # Client-side tool call (or mock)
+                parts.append(
+                    ToolCallPart(
+                        tool_name=tool_call.function.name,
+                        args=tool_call.function.arguments,
+                        tool_call_id=tool_call.id,
+                    )
+                )
+
+        # Add text content after tool calls (for server-side tools, this is the final result)
+        if response.content:
+            parts.append(TextPart(content=response.content))
+
+        # Convert usage with detailed token information
+        usage = self._map_usage(response)
+
+        # Map finish reason
+        finish_reason_map = {
+            'stop': 'stop',
+            'length': 'length',
+            'content_filter': 'content_filter',
+            'max_output_tokens': 'length',
+            'cancelled': 'error',
+            'failed': 'error',
+        }
+        raw_finish_reason = response.finish_reason
+        mapped_reason = (
+            finish_reason_map.get(raw_finish_reason, 'stop') if isinstance(raw_finish_reason, str) else 'stop'
+        )
+        finish_reason = cast(FinishReason, mapped_reason)
+
+        return ModelResponse(
+            parts=parts,
+            usage=usage,
+            model_name=self._model_name,
+            timestamp=response.created,
+            provider_name=self.system,
+            provider_response_id=response.id,
+            finish_reason=finish_reason,
+        )
+
+    def _map_usage(self, response: chat_types.Response) -> RequestUsage:
+        """Extract usage information from xAI SDK response, including reasoning tokens and cache tokens."""
+        return XaiModel.extract_usage(response)
+
+    @staticmethod
+    def extract_usage(response: chat_types.Response) -> RequestUsage:
+        """Extract usage information from xAI SDK response.
+
+        Extracts token counts and additional usage details including:
+        - reasoning_tokens: Tokens used for model reasoning/thinking
+        - cache_read_tokens: Tokens read from prompt cache
+        - server_side_tools_used: Count of server-side (built-in) tools executed
+        """
+        if not response.usage:
+            return RequestUsage()
+
+        usage_obj = response.usage
+
+        prompt_tokens = usage_obj.prompt_tokens or 0
+        completion_tokens = usage_obj.completion_tokens or 0
+
+        # Build details dict for additional usage metrics
+        details: dict[str, int] = {}
+
+        # Add reasoning tokens if available (optional attribute)
+        reasoning_tokens = getattr(usage_obj, 'reasoning_tokens', None)
```

Comment:
Do we need `getattr` here and below? If it's optional, ideally reading it when it's not provided would just give us `None`. Perhaps worth updating the xai SDK to make this a bit more natural/type-safe